File: scripts/day2/day2_scripts_Python/Day2_Application2_WHO.py
import os
import json
import pandas as pd

# =============================================================================
# Instats Seminar: Web Scraping and Data Collection
# Day 2 - From raw output to usable data
# Application 2: Health Sciences - WHO Global Health Observatory
# =============================================================================
# Goal: Clean and standardize raw JSON data into a usable tabular format.

def clean_who_data():
    """
    Reads raw cached data from Day 1, cleans it, and saves it as a CSV.
    Demonstrates handling missing values, filtering, and standardizing.
    """
    raw_cache_file = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), 
                                  'day1', 'data', 'cache', 'day1_app2_who_raw.json')
    
    clean_dir = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'data', 'cache')
    os.makedirs(clean_dir, exist_ok=True)
    clean_file = os.path.join(clean_dir, 'day2_app2_who_clean.csv')

    print(f"Loading raw data from {raw_cache_file}")
    with open(raw_cache_file, 'r', encoding='utf-8') as f:
        raw_data = json.load(f)
        
    records = raw_data.get('value', [])
    print(f"Loaded {len(records)} records. Starting cleaning process...")

    df = pd.DataFrame(records)
    
    # Select relevant columns
    cols_to_keep = ['Id', 'IndicatorCode', 'SpatialDimType', 'SpatialDim', 'TimeDimType', 'TimeDim', 'Dim1Type', 'Dim1', 'NumericValue']
    # Only keep columns that exist in the dataframe
    cols_to_keep = [c for c in cols_to_keep if c in df.columns]
    df = df[cols_to_keep]
    
    # Rename columns for better readability
    rename_map = {
        'SpatialDim': 'CountryCode',
        'TimeDim': 'Year',
        'Dim1': 'Sex',
        'NumericValue': 'LifeExpectancy'
    }
    df = df.rename(columns=rename_map)
    
    # Drop rows where LifeExpectancy is missing
    df = df.dropna(subset=['LifeExpectancy'])
    
    # Round LifeExpectancy to 1 decimal place
    df['LifeExpectancy'] = df['LifeExpectancy'].round(1)
    
    print(f"Cleaning complete. Saving {len(df)} records to {clean_file}")
    df.to_csv(clean_file, index=False)
    
    print("\nSample of cleaned data:")
    print(df.head())

if __name__ == "__main__":
    clean_who_data()

File: scripts/day2/day2_scripts_R/Day2_Application2_WHO.R
# =============================================================================
# Instats Seminar: Web Scraping and Data Collection
# Day 2 - From raw output to usable data
# Application 2: Health Sciences - WHO Global Health Observatory
# =============================================================================
# Goal: Clean and standardize raw JSON data into a usable tabular format.

library(jsonlite)
library(dplyr)
library(fs)

clean_who_data <- function() {
  raw_cache_file <- path(path_dir(path_dir(getwd())), "day1", "data", "cache", "day1_app2_who_raw.json")
  clean_dir <- path(path_dir(path_dir(getwd())), "day2", "data", "cache")
  dir_create(clean_dir)
  clean_file <- path(clean_dir, "day2_app2_who_clean.csv")
  
  cat("Loading raw data from", raw_cache_file, "\n")
  # Use simplifyVector = TRUE to get a data frame directly if possible
  raw_data <- fromJSON(raw_cache_file, simplifyVector = TRUE)
  
  df <- raw_data$value
  cat("Loaded", nrow(df), "records. Starting cleaning process...\n")
  
  # Select, rename, and filter
  clean_df <- df %>%
    select(any_of(c("Id", "IndicatorCode", "SpatialDimType", "SpatialDim", 
                    "TimeDimType", "TimeDim", "Dim1Type", "Dim1", "NumericValue"))) %>%
    rename(
      CountryCode = SpatialDim,
      Year = TimeDim,
      Sex = Dim1,
      LifeExpectancy = NumericValue
    ) %>%
    filter(!is.na(LifeExpectancy)) %>%
    mutate(LifeExpectancy = round(LifeExpectancy, 1))
  
  cat("Cleaning complete. Saving", nrow(clean_df), "records to", clean_file, "\n")
  write.csv(clean_df, clean_file, row.names = FALSE)
  
  cat("\nSample of cleaned data:\n")
  print(head(clean_df))
}

clean_who_data()
